jib-core/src/main/java/com/google/cloud/tools/jib/builder/ProgressEventDispatcher.java

```diff
@@ -0,0 +1,134 @@
+ * {@link #close}.
+ *
+ * <p>This class is <em>not</em> thread-safe. Only use a single instance per thread and create child
+ * instances with {@link #newChildProducer}.
+ */
+public class ProgressEventDispatcher implements Closeable {
+
+  /**
+   * Creates a new {@link ProgressEventDispatcher} based off an existing {@link
+   * ProgressEventDispatcher}.
+   *
+   * <p>Implementations should be thread-safe.
+   */
+  @FunctionalInterface
+  public interface Factory {
+
+    /**
+     * Creates the {@link ProgressEventDispatcher} with an associated {@link Allocation}.
+     *
+     * @param description user-facing description of what the allocation represents
+     * @param allocationUnits number of allocation units
+     * @return the new {@link ProgressEventDispatcher}
+     */
+    ProgressEventDispatcher create(String description, long allocationUnits);
+  }
+
+  /**
+   * Creates a new {@link ProgressEventDispatcher} with a root {@link Allocation}.
+   *
+   * @param eventDispatcher the {@link EventDispatcher}
+   * @param description user-facing description of what the allocation represents
+   * @param allocationUnits number of allocation units
+   * @return a new {@link ProgressEventDispatcher}
+   */
+  public static ProgressEventDispatcher newRoot(
+      EventDispatcher eventDispatcher, String description, long allocationUnits) {
+    return newProgressEventDispatcher(
+        eventDispatcher, Allocation.newRoot(description, allocationUnits));
+  }
+
+  /**
+   * Creates a new {@link ProgressEventDispatcher} and dispatches a new {@link ProgressEvent} with
+   * progress 0 for {@code allocation}.
+   *
+   * @param eventDispatcher the {@link EventDispatcher}
+   * @param allocation the {@link Allocation} to manage
+   * @return a new {@link ProgressEventDispatcher}
+   */
+  private static ProgressEventDispatcher newProgressEventDispatcher(
+      EventDispatcher eventDispatcher, Allocation allocation) {
+    return new ProgressEventDispatcher(eventDispatcher, allocation).dispatchProgress(0);
+  }
+
+  private final EventDispatcher eventDispatcher;
+  private final Allocation allocation;
+
+  private long remainingAllocationUnits;
+  private boolean closed = false;
+
+  private ProgressEventDispatcher(EventDispatcher eventDispatcher, Allocation allocation) {
+    this.eventDispatcher = eventDispatcher;
+    this.allocation = allocation;
+
+    remainingAllocationUnits = allocation.getAllocationUnits();
+  }
+
+  /**
+   * Creates a new {@link Factory} for a {@link ProgressEventDispatcher} that manages a child {@link
+   * Allocation}. Since each child {@link Allocation} accounts for 1 allocation unit of its parent,
+   * this method decrements the {@link #remainingAllocationUnits} by {@code 1}.
+   *
+   * @return a new {@link Factory}
+   */
+  public Factory newChildProducer() {
+    decrementRemainingAllocationUnits(1);
+    return (description, allocationUnits) ->
```

Comment:
Question: the name `Factory` implies that, when I have this "factory" instance, I can use it multiple times to create `ProgressEventDispatcher`s. It this allowed? Or, it is intended that the caller should call the "create()" only once? If the latter, can we prevent creating the dispatcher multiple times?

Reply:
Yes, it is intended to be called only once - will ensure this.

Reply:
Looks like this caught some mistakes!